acados/ocp_nlp/ocp_nlp_common.c

```diff
@@ -2115,11 +2242,208 @@ void ocp_nlp_embed_initial_value(ocp_nlp_config *config, ocp_nlp_dims *dims,
 }
 
 
+double ocp_nlp_compute_merit_gradient(ocp_nlp_config *config, ocp_nlp_dims *dims,
+                                  ocp_nlp_in *in, ocp_nlp_out *out, ocp_nlp_opts *opts,
+                                  ocp_nlp_memory *mem, ocp_nlp_workspace *work)
+{
+    /* computes merit function gradient at iterate: out -- using already evaluated gradients of submodules
+       with weights: work->weight_merit_fun */
+    int i, j;
+
+    int N = dims->N;
+    int *nv = dims->nv;
+    int *nx = dims->nx;
+    int *nu = dims->nu;
+    int *ni = dims->ni;
+
+    double merit_grad = 0.0;
+    double weight;
+
+    // NOTE: step is in: mem->qp_out->ux
+    struct blasfeo_dvec *tmp_vec; // size nv
+    struct blasfeo_dvec tmp_vec_nxu = work->tmp_nxu;  // size nx
+    struct blasfeo_dvec dxnext_dy = work->dxnext_dy;  // size nx
+
+    // cost
+    for (i=0; i<=N; i++)
+    {
+        tmp_vec = config->cost[i]->memory_get_grad_ptr(mem->cost[i]);
+        merit_grad += blasfeo_ddot(nv[i], tmp_vec, 0, mem->qp_out->ux + i, 0);
+    }
+    double merit_grad_cost = merit_grad;
+
+    /* dynamics */
+    double merit_grad_dyn = 0.0;
+    for (i=0; i<N; i++)
+    {
+        // get shooting node gap x_next(x_n, u_n) - x_{n+1};
+        tmp_vec = config->dynamics[i]->memory_get_fun_ptr(mem->dynamics[i]);
+
+        /* compute directional derivative of xnext with direction y -> tmp_vec_nxu */
+        blasfeo_dvecse(nx[i], 0.0, &dxnext_dy, 0);
+        for (j = 0; j < nu[i]; j++)
+        {
+            double alpha = BLASFEO_DVECEL(mem->qp_out->ux+i, j);
+            // Su
+            blasfeo_drowex(nx[i], 1.0, mem->qp_in->BAbt, j, 0, &tmp_vec_nxu, 0);
+            blasfeo_daxpy(nx[i], alpha, &tmp_vec_nxu, 0, &dxnext_dy, 0, &dxnext_dy, 0);
+        }
+        for (j = nu[i]; j < nu[i] + nx[i]; j++)
+        {
+            double alpha = BLASFEO_DVECEL(mem->qp_out->ux+i, j);
+            // Sx
+            blasfeo_drowex(nx[i], 1.0, mem->qp_in->BAbt, j, 0, &tmp_vec_nxu, 0);
+            blasfeo_daxpy(nx[i], alpha, &tmp_vec_nxu, 0, &dxnext_dy, 0, &dxnext_dy, 0);
+        }
+        /* add merit gradient contributions depending on sign of shooting gap */
+        for (j = 0; j < nx[i]; j++)
```

Comment:
also here it should be `nx[i+1]` as this is the (state) size of `pi+i` and `ux+i+1`

Reply:
Indeed I found some more related errors.
Fixed in https://github.com/acados/acados/pull/772/commits/61cd5f7a7ff691da282dab3364038e269d480ee6